### PR TITLE
* Implement setting for disabling external link opening

### LINKED ISSUE #1419: Add a way to disable "tabbing" or opening new windows
**Is your feature request related to a usage problem (not a bug)? Please describe.**
I want to spend less time on yt/the internet and that would help

**Describe the solution you'd like to see implemented**
an option to disable opening links in new windows or opening new windows entirely

**Describe alternatives you've considered**

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by PikachuEXE**
**Pull Request Type**
Please select what type of pull request this is:
- [X] Feature Implementation

**Related issue**
closes #1419

**Description**
Introduce a setting to allow disabling "opening non youtube links in browser" behaviour.
`Enable External Link Opening` (default `true`)

**Screenshots (if appropriate)**
image

**Testing (for code that is not small enough to be easily understandable)**
- Browse https://youtu.be/ECOeNxtxH0k (Win 1)
- Open another window into setting page (Win 2) & confirm new setting exists
- Click on one of the links in description part of the video in Win 1 & confirm link opened
- Update new setting in Win 2
- Click on same / different link in Win 1 & confirm no link opened

**Desktop (please complete the following information):**
 - OS: MacOS
 - OS Version: 10.15.7
 - FreeTube version: 13.1.0 (latest on `development` actually)

**Additional context**
- Not sure which setting category should this new setting entry be added to.

### CODE REVIEW

**Review by Svallinn — CHANGES_REQUESTED** [review]
Two things:
- this should be a `disable` setting rather than an `enable` one, since **turning on a distraction free setting** in this scenario would be to disable this feature, not to enable it
- a toast should be displayed to the user whenever he clicks an **external** link to inform him of this setting is turned on (aka external link opening is disabled)

**Comment by PikachuEXE** [comment]
Updated

**Comment by Svallinn** [comment]
You haven't implemented the toast message yet.

**Comment by PikachuEXE** [comment]
I totally missed that sorry 😅 
Updated now

File: src/renderer/store/modules/settings.js
Comment (Svallinn) [inline src/renderer/store/modules/settings.js:177]: There's no reason to disable it by default.
```suggestion
  disableExternalLinkOpening: false,
```

File: static/locales/en-US.yaml
Comment (Svallinn) [inline static/locales/en-US.yaml:653]:
```suggestion
External link opening has been disabled in the general settings: 'External link opening has been disabled in the general settings'
```

**Review by Svallinn — CHANGES_REQUESTED** [review]
So, I've discussed it with the rest of the team and we feel like this doesn't really belong in the distraction free settings.
We feel like it's a better fit for the general settings, so I'll have to ask you to move the logic over to the general settings (sorry for making that decision this late -_-')

**Comment by PikachuEXE** [comment]
Updated :)

I actually had no idea which setting section should this be added to and expect it might be moved.

**Comment by PikachuEXE** [comment]
Check failed due to `npm ci`
image

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
By the way u have to add the word "closes" before the issue number to link the issue. After this PR is merged the issue will close automatically instead of staying open and clutter the backlog.

**Review by peepopoggers — DISMISSED** [review]

**Comment by PikachuEXE** [comment]
@jasonhenriquez 
I am able to adjust the new setting implementation to a select
image
image

However I am not sure how to implement use `ft-prompt` to implement option for "openLinkWithPrompt".
I will continue studying the existing code for `ft-prompt` usage.

**Comment by PikachuEXE** [comment]
Implemented "open link after prompt"
I have a short recording but would be best to test it locally

https://user-images.githubusercontent.com/1018543/131947095-e6e525ea-8531-4ece-843d-5658bfd12082.mov

**Comment by jasonhenriquez** [comment]
@PikachuEXE You did an excellent job with this! It looks and feels good. Two more things:
 1. For some reason, the ft-select box is weirdly placed. You will have to fiddle around with that some.
2. The "Open Link After Prompt" text in the ft-select box is extending out into the tiny triangle. This problem has already existed here and other places for some other languages (#1300), but it is relatively infrequent for English strings. This is not a blocking issue, but I just wanted you to be aware of this.

**Comment by PikachuEXE** [comment]
@jasonhenriquez 

Placement issue fixed as the new setting was a toggle instead of a select
Moved to "select grid"

Updated "Open Link After Prompt" to become shorter 
But the root cause is that the width of select is not auto adjusted to value inside
I guess I will just use shorter value as workaround for now

Screenshot 2021-09-03 at 2 15 49 PM
Screenshot 2021-09-03 at 2 15 51 PM

**Comment by PikachuEXE** [comment]
Just fixed the style in another way: increasing the width of all select elements (in general settings section only)
What do you think?

image

**Comment by peepopoggers** [comment]
Looks Great

File: src/renderer/App.js
Comment (jasonhenriquez) [inline src/renderer/App.js:317]: Make a computed property for getExternalLinkHandling and use it here.

File: src/renderer/App.js
Comment (jasonhenriquez) [inline src/renderer/App.js:322]: See above comment.

**Review by jasonhenriquez — CHANGES_REQUESTED** [review]
I haven't tested yet, but just one more minor suggestion. Spectacular job @PikachuEXE!

**Comment by PikachuEXE** [comment]
Code change made:
image

Unexpected result:
image

**Comment by jasonhenriquez** [comment]
@ line 320: It's called `externalLinkHandling`, not `getExternalLinkHandling`.

**Comment by PikachuEXE** [comment]
Done
I have no idea why I can't spot that obvious mistake back then

**Comment by jasonhenriquez** [comment]
It looks good and works well. I do have one final question, though, which is whether it is possible in FreeTube for people to put custom text for a URL. If not, I have no further suggestions. If so, the actual link should be displayed in the toast (when external link opening is disabled), the modal (when external link opening is set to prompt), and as a title on hover (in any case). But again, this may not be the case.

File: src/renderer/App.js
Comment (jasonhenriquez) [reply src/renderer/App.js:326]: These still need to be addressed as well.

File: src/renderer/components/general-settings/general-settings.js
Comment (ChunkyProgrammer) [inline src/renderer/components/general-settings/general-settings.js:168]: Bit of nitpick but I think "Don't Open Link" might be better text than "Do Nothing" (especially when considering translators might be missing the context of the string)

Comment (PikachuEXE) [reply src/renderer/components/general-settings/general-settings.js:168]: I found it strange to read "not doing specific thing" as an "action to be performed by the application".
I can consider changing it to "No Action" if that sounds better to you.

**Comment by PikachuEXE** [comment]
Screenshot for latest change about showing URL in prompt
image

File: src/renderer/components/general-settings/general-settings.js
Comment (ChunkyProgrammer) [reply src/renderer/components/general-settings/general-settings.js:168]: No Action would be better

Comment (PikachuEXE) [reply src/renderer/components/general-settings/general-settings.js]: Updated to become "No Action"

**Comment by PikachuEXE** [comment]
@jasonhenriquez 
> whether it is possible in FreeTube for people to put custom text for a URL

I have no idea what you are asking

**Comment by jasonhenriquez** [comment]
> @jasonhenriquez
> 
> > whether it is possible in FreeTube for people to put custom text for a URL
> 
> I have no idea what you are asking

[Something like this.](https://duckduckgo.com) Where the text is not https://duckduckgo.com but the link is.

**Comment by PikachuEXE** [comment]
OK I guess I know what you mean
According to https://www.tubics.com/blog/youtube-description-text/
Only URLs are converted to links when displayed. So there should be no link with custom text.

**Comment by peepopoggers** [comment]
Is this ready to be implimented?

**Comment by jasonhenriquez** [comment]
My requested changes from earlier are all to my knowledge.

**Comment by PikachuEXE** [comment]
> Is this ready to be implemented?

Do you mean is it ready to be merged?

**Comment by peepopoggers** [comment]
@PikachuEXE yes

**Comment by PikachuEXE** [comment]
I think I have addresses all change requests

**Comment by jasonhenriquez** [comment]
> I think I have addresses all change requests

You can check the PR and see it: lines 321 and 326. Perhaps the change addressing that has gotten removed in one of the force-pushes. In the future, I do advise that you avoid changing history whenever possible. It makes things harder to track, even if it can seem messier.

**Comment by PikachuEXE** [comment]
I search the whole history again and found that 2 comments hidden for some reason.
I will update accordingly soon.

File: static/locales/en-US.yaml
Comment (jasonhenriquez) [inline static/locales/en-US.yaml:153]: behavior* for American English. Sorry, another final nitpick.

Comment (jasonhenriquez) [reply static/locales/en-US.yaml:153]: I'm also not sure if "non YouTube url" would translate well. I think something like "a link that opens in your default browser" would be more translatable.

Comment (PikachuEXE) [reply static/locales/en-US.yaml:153]: > `behavior`

I can never remember which one is US and which one is UK variant @@

Updated text to 
>
        External Link Handling: Choose the behavior when a link is clicked.
        By default FreeTube will open the link in your default browser.
        (Unless it points to a Youtube Video/Playlist/Channel etc.)
        You can choose to disable it completely for safety.

I think this setting is controlling **any** link except those pointing to Youtube "resources".
Though I am not sure if this updated text can be easily comprehended and translated.

File: static/locales/en-US.yaml
Comment (jasonhenriquez) [inline static/locales/en-US.yaml:156]: 1. This needs to be under the Tooltips section and follow the same conventions as it does there.
2. The "unless" clause again seems difficult to translate the way it is now grammatically.
3. This explanation does not talk about the prompting option at all.

My suggestion:
Choose the default behavior for interacting with links to content that cannot be opened in FreeTube.

File: static/locales/en-US.yaml
Comment (jasonhenriquez) [inline static/locales/en-US.yaml:158]: I'm thinking "Open After Prompt" is somewhat unclear and suggests that it will inevitably open the link. I recommend "Ask Every Time" instead.

**Review by jasonhenriquez — CHANGES_REQUESTED** [review]

File: static/locales/en-US.yaml
Comment (PikachuEXE) [reply static/locales/en-US.yaml]: Translation entry moved and content updated

File: static/locales/en-US.yaml
Comment (PikachuEXE) [reply static/locales/en-US.yaml]: Updated to "Ask Before Opening Link"

**Review by jasonhenriquez — APPROVED** [review]
I approve. @PikachuEXE Thank you for being so incredibly patient and willing to make changes after my innumerable nitpicks.

**Comment by PikachuEXE** [comment]
This took a while but I think it's worth the effort and time
@jasonhenriquez Thank **you** for doing a review in detail!

@peepopoggers This PR is Ready to be merged now


Strange, now I forgot why I started implementing this feature.
I don't need this feature myself o_O

**Review by efb4f5ff-1298-471a-8973-3d47447115dc — APPROVED** [review]
Nicely done👍👍👍

**Review by peepopoggers — APPROVED** [review]

**Comment by efb4f5ff-1298-471a-8973-3d47447115dc** [comment]
Merge is being blocked? There is are changes requested by someone but im failing to see what request is blocking this.

**Comment by jasonhenriquez** [comment]
> Merge is being blocked? There is are changes requested by someone but im failing to see what request is blocking this.

I believe @Svallinn has to dismiss or change their review for the merge to occur.

**Pull request merged**
